Support for searches with logical operators and $maxDistance

etsuo opened the pull request:
Fix #271

Requires PR: https://github.com/strongloop/loopback-datasource-juggler/pull/1036

Updated `buildNearFilter` to account for changes to juggler's nearFilter. Added support for logical operators and $maxDistance.

0candy commented:
Thank you for submitting the PR. Please include a unit test to verify your changes.

etsuo commented:
@0candy PR updated w/ unit tests and shortened commit messages.

etsuo commented:
Moved this comment out of the PR message in an attempt to satisfy the commit linter: 

Requires PR: https://github.com/strongloop/loopback-datasource-juggler/pull/1028

Updated `buildNearFilter` to account for changes to juggler's nearFilter. Added support for logical operators and $maxDistance.

etsuo commented:
@0candy is the linter complaining about my PR message at the top... or is it expecting me to alter the commit message on the actual commit with a rebase? It's a bit confusing because the first commit with the really detailed commit message has a green check. The two subsequent commits are complaining... but they have really short commit messages. I feel like I'm trying to comply with some professor's obsession with the Chicago Manual of Style... :p

etsuo commented:
@0candy I finally figured out what the commit linter wants- I mistakenly thought it was the fist comment on this PR that it was complaining about... I squashed my commits and reworded to something more abbreviated.

0candy commented on lib/mongodb.js:
```diff
@@ -734,8 +734,13 @@ MongoDB.prototype.buildSort = function(model, order) {
+  near.key.forEach(function(key) {
+    node = node[key];
+  });
+
```
Did you mean to have the below code in this loop? Otherwise, what's the point of looping through the keys?

jannyHou replied:
Same question ^

etsuo replied:
The idea here is that `var node` holds a reference starting at the top of the `query` object with `query.where`. Each `key` (generated by the `searchForNear` function in juggler's `geo.js`) walks the reference down a "node" in the query object until it drills down to the key that has the location. Let me know if that still isn't clear or if there's a way I could code it so that it's more readable.

0candy replied:
```+  near.key.forEach(function(key) {
+    node = node[key];
+  });```

Your code here will loop through all the keys and node will be left assigned to the last key it encountered.

jannyHou replied:
And please notice that due to the check in loopback-datasource-juggler: https://github.com/strongloop/loopback-datasource-juggler/blob/839bb9a81eac236f7356651881def0c67088e007/lib/dao.js#L1793-L1796
function `buildNearFilter` will NOT be called when it's with logic operators. 
Because `geo.nearFilter(query.where);` doesn't check operators.

0candy commented:
@Amir-61 @superkhau @jannyHou PTAL

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+          createPost.bind(null),
+          createPost.bind(null),
+        ], function(err) {
+          should.not.exist(err);
```
Let's refactor the error assertion as `if (err) return done(err)`, example: https://github.com/strongloop/loopback-datasource-juggler/blob/e3b6b7891c009354f36f7f811428149c61439c71/test/manipulation.test.js#L287

> should.not.exist(err)` fails the test with something like "expected Error to not exist", but callback error will enable the test runner to show the original error message and stack trace.

etsuo replied:
Sure - no problem. I'll do that right now. I was suspicious of this pattern, but was trying to stick with the pattern established by the existing geo tests.

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+              ],
+            },
+          }, function(err, results) {
+            should.not.exist(err);
```
ditto `if (err) return done(err)`, thanks!

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+            }, callback);
+          },
+        ], function(err) {
+          should.not.exist(err);
```
ditto if (err) return done(err), thanks!

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+                  ],
+                },
+              }, function(err, results) {
+                should.not.exist(err);
```
ditto if (err) return done(err), thanks!

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+                  },
+                },
+              }, function(err, results) {
+                should.not.exist(err);
```
ditto if (err) return done(err), thanks!

jannyHou commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+              });
+            },
+          ], function(err) {
+            should.not.exist(err);
```
ditto if (err) return done(err), thanks!

etsuo commented:
@jannyHou @0candy refactors applied.

superkhau commented on lib/mongodb.js:
```diff
@@ -744,26 +749,42 @@ function buildNearFilter(query, near) {
+    if (node && typeof node === 'object') {
+      var keys = Object.keys(node);
+      for (var i = 0; i < keys.length; i++) {
+        var ex = node[keys[i]];
```
what is `ex`, please use a more descriptive name

superkhau commented on lib/mongodb.js:
```diff
@@ -744,26 +749,42 @@ function buildNearFilter(query, near) {
+      for (var i = 0; i < keys.length; i++) {
+        var ex = node[keys[i]];
+        keyPath.push(keys[i]);
+        if (ex && ex.near) {
```
ditto

etsuo replied:
@superkhau I inherited this naming convention and wasn't sure how you guys feel about PRs changing the names of variables... I'll submit an update with something that makes more sense to me. If you have a suggestion ahead of that, please let me know. :)

superkhau replied:
Where did you inherit it from? Something that makes more sense is fine.

etsuo replied:
@superkhau sorry, "inherited" was probably not clear in this context... I meant that the code was written as such by the prior contributor and I simply augmented it with the intention of not changing anything that was relevant to the fix. I updated the variable name to `nodeProperty` - let me know if that's still not clear. Thanks.

superkhau commented:
Parts of this PR look similar to https://github.com/strongloop/loopback-connector-mongodb/pull/259/files. @0candy @Amir-61 @jannyHou PTAL and ping me again after.

0candy commented:
@superkhau Yes, there is a bit of overlap. The other PR fixes max and min distance. This one fixes near filter and max distance. So the overlap is in the max distance.

etsuo commented:
@superkhau @0candy - the `maxDistance` fix resolved a blocking issue internally on a project I'm working on, so I threw it in. It's three lines of code, I can remove it if you'd like.

etsuo commented:
@superkhau the puzzling and opaque `ex` has been renamed. ;)

superkhau commented:
@0candy Can you review this? There are some obvious things to fix and you're on triage this week. I'll take a look later once you've left your opinions.

Tuxity commented on test/mongodb.test.js:
```diff
@@ -1438,6 +1439,156 @@ describe('mongodb connector', function() {
+                    {
+                      location: {
+                        near: new GeoPoint(coords),
+                        maxDistance: Number(Math.round(5 * 1609.34).toFixed(2)), // miles to meters
```
There is a `unit` keyword avalaible https://docs.strongloop.com/display/public/LB/Where+filter#Wherefilter-near(orderingandlimitingbydistance)

0candy commented:
@etsuo Not sure if you saw my comment above:
https://github.com/strongloop/loopback-connector-mongodb/pull/273#r74457718

slnode commented:
Can one of the admins verify this patch?

superkhau commented:
FYI, #259 was merged in so you might need to make some changes to remove duplicate efforts.

jannyHou commented:
Hi @etsuo any update on applying https://github.com/strongloop/loopback-connector-mongodb/pull/273#r74457718?

slnode commented:
Can one of the admins verify this patch?

ebarault commented:
please, can you check if these PR fixes your issue?
see strongloop/loopback-datasource-juggler#1216
and #322

ebarault commented:
hi, could you check if these PR fix your issue ?
strongloop/loopback-datasource-juggler#1216
strongloop/loopback-connector-mongodb#322
